irods/test/connection_test.py

```diff
@@ -65,6 +67,47 @@ def test_reply_failure(self):
             with self.assertRaises(NetworkException):
                 conn.reply(0)
 
+    def test_that_connection_timeout_works__issue_377(self):
+        sess = self.sess
+
+        # skip if server is remote
+        if sess.host not in ('localhost', socket.gethostname()):
+            self.skipTest('Requires access to server-side file(s)')
+        r = sess.resources.create('hello', 'unixfilesystem', host='localhost')
+        rand = os.urandom(1024)*64
+
+        # Create a large file.
+        size = 1024**3*4
+        physical_path = '/tmp/testfile'
+        with open(physical_path,'wb') as f:
+            while f.tell() < size:
+                f.write(rand)
+
+        # Set a very short socket timeout and remove all pre-existing socket connections.
+        # This forces a new connection to be made for any ensuing connections to the iRODS server.
+        sess.connection_timeout = 1.0
+        sess.cleanup()
+
+        # Perform a time-consuming operation in the server (registering a new object such that the server
+        # computes its checksum) during which the socket will time out.
+        coll = sess.collections.create(helpers.home_collection(sess) + '/testdir_377')
+        try:
+            with self.assertRaises(NetworkException):
+                sess.data_objects.register(physical_path, coll.path+'/testfile',
+                                           **{kw.REG_CHKSUM_KW: '', kw.RESC_NAME_KW: r.name})
+        finally:
+            # Set the connection pool's socket timeout interval back to default, and clean up.
+            sess.connection_timeout = None
+            sess.cleanup()
+            coll.remove(recurse = True, force = True)
+            r.remove()
+            if os.path.isfile(physical_path):
+                os.unlink(physical_path)
+
+    def test_connection_timeout_parameter_in_session_init__issue_377(self):
+            timeout_value = 1.0
+            s = helpers.make_session(connection_timeout = timeout_value)
+            self.assertEqual(s.pool.connection_timeout, timeout_value)
```

Comment:
Maybe worth adding a section (or a separate test) which tests the `None` case?

Reply:
> Maybe worth adding a section (or a separate test) which tests the `None` case?

Possibly, yes.  I have some doubt about the present treatment of defaults, so that might change.   I have now come to the belief that
   - session constructor should default the connection_timeout parameter to the value of `irods.DEFAULT_CONNECTION_TIMEOUT = 120`
   - `None`, like any other value given as the value of this parameter, should filter down as the pool value of `connection_timeout`, and in the case of `None`, it will mean that the socket behaves according to the OS default (as determined by `socket.getdefaulttimeout( )`

Reply:
Done, at least to the extent that we know the timeout value requested does propagate down to the socket object, and it is verified in a test.